jstz_api/src/urlpattern/mod.rs

```diff
@@ -0,0 +1,576 @@
+//!
+//! More information:
+//!  - [MDN documentation][mdn]
+//!  - [WHATWG `URLPattern` specification][spec]
+//!
+//! [mdn]: https://developer.mozilla.org/en-US/docs/Web/API/URLPattern
+//! [spec]: https://urlpattern.spec.whatwg.org/
+
+use boa_engine::{
+    js_string,
+    object::{builtins::JsArray, Object},
+    property::Attribute,
+    value::TryFromJs,
+    Context, JsArgs, JsError, JsNativeError, JsObject, JsResult, JsString, JsValue,
+    NativeFunction,
+};
+use boa_gc::{custom_trace, Finalize, GcRefMut, Trace};
+use jstz_core::{
+    accessor,
+    native::{
+        register_global_class, Accessor, ClassBuilder, JsNativeObject,
+        JsNativeObjectToString, NativeClass,
+    },
+    value::IntoJs,
+};
+use urlpattern::quirks::StringOrInit as InnerStringOrQuirksInit;
+use urlpattern::quirks::UrlPatternInit as InnerUrlPatternQuirksInit;
+use urlpattern::UrlPattern as InnerUrlPattern;
+use urlpattern::UrlPatternComponentResult as InnerUrlPatternComponentResult;
+use urlpattern::UrlPatternResult as InnerUrlPatternResult;
+
+pub struct UrlPatternInput(InnerStringOrQuirksInit);
+impl Default for UrlPatternInput {
+    fn default() -> Self {
+        Self(InnerStringOrQuirksInit::Init(
+            InnerUrlPatternQuirksInit::default(),
+        ))
+    }
+}
+#[derive(Default)]
+pub struct UrlPatternInit(InnerUrlPatternQuirksInit);
+
+pub struct UrlPatternComponentResult(InnerUrlPatternComponentResult);
+pub struct UrlPatternResult {
+    // It should be UrlPatternInit instead of UrlPatternInput
+    // according to Deno types?
+    pub(crate) inputs: Vec<UrlPatternInput>,
+    pub(crate) url_pattern_result: InnerUrlPatternResult,
+}
+
+#[derive(Finalize)]
+pub struct UrlPattern {
+    pub(crate) url_pattern: InnerUrlPattern,
+}
+
+unsafe impl Trace for UrlPattern {
+    custom_trace!(_this, {});
+}
+
+impl JsNativeObjectToString for UrlPattern {
+    fn to_string(
+        this: &JsNativeObject<Self>,
+        context: &mut Context<'_>,
+    ) -> JsResult<JsValue> {
+        let s = format!("{:?}", this.deref().url_pattern);
+        Ok(s.into_js(context))
+    }
+}
+
+impl UrlPattern {
+    fn process_input(
+        string_or_init: InnerStringOrQuirksInit,
+        base_url: Option<String>,
+    ) -> JsResult<(
+        urlpattern::UrlPatternMatchInput,
+        (InnerStringOrQuirksInit, Option<String>),
+    )> {
+        match urlpattern::quirks::process_match_input(string_or_init, base_url.as_deref())
+        {
+            Err(e) => Err(JsError::from_native(
+                JsNativeError::typ().with_message(e.to_string()),
+            )),
+            Ok(Some(input)) => Ok(input),
+            Ok(None) => Err(JsError::from_native(JsNativeError::error())),
+        }
+    }
+
+    // We do not support options (ignoreCase), as it is not supported in Deno
+    // nor in `urlpattern` crate. There is an open PR for supporting it in
+    // `urlpattern`, and we could use it when it gets merged.
```

Comment:
Add a `FIXME` which references a GitHub issue in `jstz` for this

Reply:
Done, added also a FIXME on the top also saying that we still don't support the whole spec (I saw this on other APIs).